contributors/design-proposals/hardware-accelerators.md

```diff
@@ -0,0 +1,133 @@
+# Hardware Accelerators
+
+Author: Vishnu Kannan (vishh@)
+
+## Introduction
+
+Hardware Accelerators are becoming a widely used commodity across various industries.
+Accelerators can bring down computing latency and/or costs significantly.
+Many of these accelerators present unique technical constraints.
+Kubernetes as an established Application Management Platform can play a crucial role in getting workloads that require hardware acceleration to be portable and function at scale.
+Most of the large public clouds already provide various types of Hardware Accelerators.
+There are quite a few Kubernetes users who are already managing Accelerators at scale on their own datacenters.
+
+The remainder of this document provides a technical overview of how hardware accelerators will be integrated into Kubernetes.
+
+## Goals
+
+* Make Kubernetes Machine Learning and Deep Learning friendly.
+* Extensible support for various types of hardware accelerators Nvidia GPUs, AMD GPUs and Google Tensor Processing Units
+* Portability across Kubernetes clusters
+* Limit feature creep to Kubernetes nucleus
+* Provide baseline application performance guarantees
+
+## Non Goals
+* Support for Cloud Gaming, Simulations, Remote Desktops and other workloads
+  * Support for these workloads will be tackled once support for Machine Learning matures
+
+## System Design
+
+The following sections highlight some of the critical design points for supporting Hardware Accelerators
+
+### API
+
+A plethora of Hardware accelerators exist in the world.
+Some of them are general purpose, but many of them are purpose built for specific use cases.
+Exposing all hardware accelerators as well known (first class) Compute Resource Types will bloat the API and compromise portability.
+For this reason, Hardware Accelerators are expected to be handled as “Extended Compute Resources”.
+
+Kubernetes nucleus will recommend and document a general purpose resource name for each family of accelerators - examples include `nvidia-gpu`, `amd-gpu`, `google-tpu`, etc., with a standard prefix `extensions.kubernetes.io`. This naming scheme partially mimics PCI ID - `<Vendor Name>-<Device Type>`.
```

Comment:
@jiayingz  there are at least a few options available there:

1. Encode whether the resource can be overcommitted somehow in the resource name.
1. Maintain a list of either overcommit or no-overcommit resource names and use them in admission control.
1. Don't allow overcommit for now, except for first-class resources. A future API for resources could allow enabling overcommit on a case-by-case basis.
    1. Do this in a way that soft-breaks OIR (currently they can be overcommitted)
    1. Special-case OIR to preserve their overcommit behavior.

(1) seems likely to require a future breaking change.
(2) potentially couples admission to extended resource names.
(3.a) might be OK but we don't have a good way to determine the extent of the breakage across the ecosystem so I would vote not to do this without a deprecation cycle.
(3.b) could be a decent stopgap, made more tolerable if we deprecate OIR at the same time. The breakage due to feature removal would be that `pod.alpha.kubernetes.io/opaque-int-<foo>` resources can no longer be overcommited.